### PR TITLE
Log when events are (unexpectedly) filtered out of responses in tests

### PR DESCRIPTION
**Pull request by MadLittleMods**
Log when events are (unexpectedly) filtered out of responses in tests.

See https://github.com/matrix-org/synapse/pull/14095#discussion_r990335492

It would be nice to have these logs as a default in CI and if changing the log level locally, having to choose between useful `INFO` and noisy `DEBUG` is not great. I'm thinking about having these logs shown by default 🤔. It probably doesn't make sense to have these show at the `INFO` level for production but I think it would make sense to show them in tests.

I'm also conscious of this being noise itself but I feel like these logs are one of the easy things to eliminate a bunch of possibilities. For example, if I see that a relevant event is an `outlier`, `soft-failed`, then that's a good unexpected indicator explaining it. `filter_events_for_client` is used in `/sync`, `/messages`, `/context` which are all common end-to-end assertion touch points (also notifications, relations).

I'm happy to close this PR if you're not seeing it 🙂 (the idea of it might be better than doing this)

### Dev notes

```
SYNAPSE_POSTGRES=1 SYNAPSE_POSTGRES_USER=postgres SYNAPSE_TEST_LOG_LEVEL=INFO python -m twisted.trial tests.test_visibility.FilterEventsForClientTestCase
```

### Todo

 - [x] Adjust logging configs for tests to show these logs by default

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by MadLittleMods** [comment]
@clokep I think it's worth trying. Will try to move it forward when I get backend-facing again. If you want to move it forward, feel free.

File: docker/conf/log.config [hunk 1]
```diff
@@ -49,17 +49,35 @@ handlers:
...
-  If SYNAPSE_LOG_SENSITIVE is unset, then override synapse.storage.SQL to INFO
-  so that DEBUG entries (containing sensitive information) are not emitted.
-#}
 loggers:
+    # This is just here so we can leave `loggers` in the config regardless of whether
+    # we configure other loggers below (avoid empty yaml dict error).
+    _placeholder:
+        level: "INFO"
+
+    {% if not SYNAPSE_LOG_SENSITIVE %}
```
Comment (MadLittleMods) [inline docker/conf/log.config:58]: Should we remove `SYNAPSE_LOG_SENSITIVE` in favor of `SYNAPSE_LOG_TESTING` here?

The docs around `SYNAPSE_LOG_SENSITIVE` already mention it's for "a developer attempting to debug something
  particularly tricky."

File: tests/test_utils/logging_setup.py [hunk 2]
```diff
@@ -54,4 +54,10 @@ def setup_logging() -> None:
     log_level = os.environ.get("SYNAPSE_TEST_LOG_LEVEL", "ERROR")
     root_logger.setLevel(log_level)
 
+    # Log when events are (maybe unexpectedly) filtered out of responses in tests. It's
+    # just nice to be able to look at the CI log and figure out why an event isn't being
+    # returned.
+    logging.getLogger("synapse.visibility.filtered_event_debug").setLevel(logging.DEBUG)
+
+    # Blow away the pyo3-log cache so that it reloads the configuration.
```
Comment (MadLittleMods) [inline tests/test_utils/logging_setup.py:62]: This comment is just copied from 

https://github.com/matrix-org/synapse/blob/703a8f9c67cfe25b956dfdcca654818d52fa7ebd/synapse/config/logger.py#L300-L301

File: tests/test_utils/logging_setup.py [hunk 3]
```diff
@@ -54,4 +54,10 @@ def setup_logging() -> None:
     log_level = os.environ.get("SYNAPSE_TEST_LOG_LEVEL", "ERROR")
     root_logger.setLevel(log_level)
 
+    # Log when events are (maybe unexpectedly) filtered out of responses in tests. It's
+    # just nice to be able to look at the CI log and figure out why an event isn't being
+    # returned.
+    logging.getLogger("synapse.visibility.filtered_event_debug").setLevel(logging.DEBUG)
```
Comment (clokep) [inline tests/test_utils/logging_setup.py:60]: I don't feel strongly, but i find it a little odd to enable this for default in tests when we seem to go to great lengths to not log anything?

**Review by clokep — COMMENTED** [review]
Seems reasonable overall.

File: tests/test_utils/logging_setup.py (hunk 3 above)
Comment (MadLittleMods) [reply tests/test_utils/logging_setup.py:60]: > we seem to go to great lengths to not log anything?

Is this based off the fact that we only log at the `ERROR` level by default when running trial tests?

hmm, I am used to always using `SYNAPSE_TEST_LOG_LEVEL=INFO` in my trial test runs so I didn't consider this. Although we do use `INFO` level for Complement test runs by default.

Perhaps based on this, we don't want it by default for trial tests but would be fine for Complement. I was originally was wanting this while investigating things in a Complement context anyway.

Comment (clokep) [reply tests/test_utils/logging_setup.py:60]: > > we seem to go to great lengths to not log anything?
> 
> Is this based off the fact that we only log at the `ERROR` level by default when running trial tests?

Mostly, yes.

Comment (clokep) [reply tests/test_utils/logging_setup.py:60]: Maybe turn it on automatically if the logger is set to INFO or DEBUG?

**Review by clokep — APPROVED** [review]
Happy if CI is happy!

**Comment by MadLittleMods** [comment]
Thanks for the review @clokep 🦙

**Pull request merged**